Link check fixes

stegru opened the pull request:
* Forwards the browser's `user-agent` header to the site being checked (some sites reject requests without one)
* ~~`405 Method not allowed` status will be assumed as the link being valid.~~
* A `GET` request will be sent if `HEAD` fails.

christopher-rtf reviewed (commented):
Why are we doing this, in particular?
'Assuming response "405 Method not allowed" means the link is working.'

I ask because this doesn't mean that "GET" is allowed.  It just means that the server unfortunately doesn't understand "HEAD".  Shouldn't we be retrying with GET instead?

What am I missing?

christopher-rtf commented:
Also: why is the API server doing these checks?  Is there a technical reason why the browser can't do them?  

Also: does our current method create any potential for us to become an arbitrary redirect endpoint?  Can it cause us to download illegal or potentially harmful content to our server, to become a vector for DDoS, etc.?

stegru commented:
> I ask because this doesn't mean that "GET" is allowed. It just means that the server unfortunately doesn't understand "HEAD". Shouldn't we be retrying with GET instead?

Yes. I added that special case before I implemented the GET retry.

christopher-rtf commented:
> > I ask because this doesn't mean that "GET" is allowed. It just means that the server unfortunately doesn't understand "HEAD". Shouldn't we be retrying with GET instead?
> 
> Yes. I added that special case before I implemented the GET retry.

Okay, let's go ahead and remove the "HEAD returning 405 is success" code then (if you haven't already done so).  Thank you!

stegru commented:
> Also: why is the API server doing these checks? Is there a technical reason why the browser can't do them?

Yes, the browser does not allow cross-origin requests, unless the server allows it with CORS headers.

> Also: does our current method create any potential for us to become an arbitrary redirect endpoint? Can it cause us to download illegal or potentially harmful content to our server, to become a vector for DDoS, etc.?

Yes... the comment fixes this.

```
    /// This allows the web client to check if a link is working.
    /// The browser is unable to perform this check itself, because cross-origin requests are forbidden
    /// in https.
    ///
    /// This does mean the server is performing network requests on behalf of the user, so there is a potential
    /// for trouble with accessing random resources.
    ///
    /// Risks:
    /// - The server is used to access something due to IP filtering.
    /// - An action on a remote server could be performed under the guise of the server.
    ///
    /// There are some limitations and mitigations:
    /// - Only authenticated users can access this.
    /// - This endpoint is accessed via a POST request.
    /// - Nothing is actually read, stored, or parsed - only the HTTP response headers.
    /// - Only normal looking URLs are accepted (http/https, non-IP, standard port)
    /// - HEAD requests are sent, then retried with a GET (some servers don't support HEAD).
    /// - The response from this endpoint is only success or failure (or validation error), nothing from
    ///    the external link is provided.
    /// - X-Forwarded-For provides the client's IP.
```
